include/deal.II/non_matching/mapping_info.h

```diff
@@ -157,30 +293,550 @@ namespace NonMatching
+    unsigned int n_faces    = 0;
+    unsigned int cell_index = 0;
+    for (const auto &cell : cell_iterator_range)
+      {
+        cell_index_offset[cell_index] = n_faces;
+        n_faces += cell->n_faces();
+        ++cell_index;
+      }
+
+    // fill unit points index offset vector
+    unit_points_index.resize(n_faces + 1);
+    cell_index                 = 0;
+    unsigned int n_unit_points = 0;
+    for (const auto &cell : cell_iterator_range)
+      {
+        for (const auto &f : cell->face_indices())
+          {
+            const unsigned int current_face_index =
+              cell_index_offset[cell_index] + f;
+
+            unit_points_index[current_face_index] = n_unit_points;
+            n_unit_points +=
+              quadrature_vector[cell_index][f].get_points().size();
+          }
+
+        ++cell_index;
+      }
+    unit_points_index[n_faces] = n_unit_points;
+
+    // compress indices
+    if (do_cell_index_compression)
+      cell_index_to_compressed_cell_index.resize(n_unfiltered_cells,
+                                                 numbers::invalid_unsigned_int);
+
+    // fill unit points and mapping data for every face of all cells
+    unit_points.resize(n_unit_points);
+    mapping_data.resize(n_faces);
+    cell_index = 0;
+    QProjector<dim> q_projector;
+    for (const auto &cell : cell_iterator_range)
+      {
+        const auto &quadratures_on_faces = quadrature_vector[cell_index];
+
+        Assert(quadratures_on_faces.size() == cell->n_faces(),
+               ExcDimensionMismatch(quadratures_on_faces.size(),
+                                    cell->n_faces()));
+
+        for (const auto &f : cell->face_indices())
+          {
+            const auto &quadrature_on_face = quadratures_on_faces[f];
+
+            const auto quadrature_on_cell =
+              q_projector.project_to_face(cell->reference_cell(),
+                                          quadrature_on_face,
+                                          f);
+
+            const auto &unit_points_on_cell = quadrature_on_cell.get_points();
+
+            const unsigned int current_face_index =
+              cell_index_offset[cell_index] + f;
+
+            auto it =
+              unit_points.begin() + unit_points_index[current_face_index];
+            for (const auto &unit_point : unit_points_on_cell)
+              {
+                *it = unit_point;
+                ++it;
+              }
+
+            compute_mapping_data_for_face_quadrature(
+              cell, f, quadrature_on_face, mapping_data[current_face_index]);
+          }
+        if (do_cell_index_compression)
+          cell_index_to_compressed_cell_index[cell->active_cell_index()] =
+            cell_index;
+
+        ++cell_index;
+      }
+
+    state = State::faces_on_cells_in_vector;
   }
 
 
 
   template <int dim, int spacedim>
-  const std::vector<Point<dim>> &
-  MappingInfo<dim, spacedim>::get_unit_points() const
+  const std::vector<Point<dim>>
```

Comment:
I think we discussed this in person already, but I am not happy about this code change because we might create (many) copies as a return value. I think it would be better if we could return an `ArrayView<Point<dim>>` here. It breaks the interface, but so does the change you do, but I think the break is the least intrusive way while still ensuring good performance.

Reply:
Done. I see some slight decay in the performance of `FEPointEvaluation` comparing this branch with master (even after this change). Do you think this is the reason for this?

Reply:
Possibly yes. Memory allocations of many small arrays in inner loops are one of worst performance sinks. As said in the other post, we do not necessarily allocate memory in the assignment to a vector, but as far as I understand the code I think it needs to happen here, because we first must create a temporary `std::vector` object to hold the temporary data. Return-value optimization, which is an often-said argument why pass-by-value should be OK, is likely not possible because we do not simply return a variable stored internally. Thus, I suggest we change this here and see how many places it affects in the tests of deal.II, which might give us an indication of breakage of user codes.